---
title: "Integrate information and categorize variants"
author: "Ji-Dung Luo"
created: "2019-08-12"
revised: "2023-07-25"
date: "2023-07-25, revised"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Set criteria
* Select the variants with
  + total depth >= 5
  + counts of minor allele >= 2
  + identified in at least three samples
* Input files
  + *hg19_UCSC_Alu_env.RData*: The Alu elements of hg19 in GRange object. (can be found in this GitHub repository)
  + The RData containing matrix of DP/AC/AF, generated in previous step.
```{r set_env,include=TRUE,eval=FALSE}
# Set cut-off
dep_cut <- 5 # total depth >=5
ac_cut <- 2 # alleleic counts >= 2
num_cut <- 3 # Identified in at least 3 samples
#
load("hg19_UCSC_Alu_env.RData") # Alu fragments in hg19, with GRange format
load("Path to varints RData")
```

## Extract information of variant annotation
```{r anno_ext,include=TRUE,eval=FALSE}
base_dir <- "path to the main directory"
res_dir <- "path to the result directory"
dir.create(res_dir)
#
annoTXT <- file.path(base_dir,"annoTab.txt")
annoTab <- read.table(annoTXT,header = FALSE,sep="\t",stringsAsFactors = FALSE)
names(annoTab) <- c("variant","geneSymbol","SNP138","geneFunc","exonFunc","AAChange","Description")
annoTab <- dplyr::select(annoTab,variant,geneSymbol,SNP138,geneFunc,exonFunc,AAChange,Description)
annoTab$geneFunc[grepl(annoTab$geneFunc,pattern = "ncRNA")] <- "ncRNA"
annoTab$geneFunc[grepl(annoTab$geneFunc,pattern = "downstream")|
                   grepl(annoTab$geneFunc,pattern = "upstream")] <- "intergenic"
annoTab$geneFunc[grepl(annoTab$geneFunc,pattern = "UTR5;UTR3")] <- "UTR3"
annoTab$geneFunc[grepl(annoTab$geneFunc,pattern = "exonic;splicing")] <- "exonic"
annoTab$geneFunc[grepl(annoTab$geneFunc,pattern = "exonic")] <- "CDS"
```

## Integrate DP/AC/AF matrix with annotation table
```{r int_info_var,include=TRUE,eval=FALSE}
# depth of each sample
varDP <- data.frame(variant=rownames(matDP))
varDP <- cbind(varDP,matDP)
names(varDP) <- c("variant",paste0("DP_",names(varDP)[-1]))
varDP$variant <- as.vector(varDP$variant)
#
# alleleic frequency (AF) for each sample
varAF <- data.frame(variant=rownames(matAF))
varAF <- cbind(varAF,matAF)
names(varAF) <- c("variant",paste0("AF_",names(varAF)[-1]))
varAF$variant <- as.vector(varAF$variant)
#
# alleleic counts (AC) for each samplle
varAC <- data.frame(variant=rownames(matAC))
varAC <- cbind(varAC,matAC)
names(varAC) <- c("variant",paste0("AC_",names(varAC)[-1]))
varAC$variant <- as.vector(varAC$variant)
#
# Merge DP, AF, and AC into single table
bigTab <- merge(varDP,varAF,id='variant')
bigTab <- merge(bigTab,varAC,id='variant')
```

## Additional information
* Add information
  + var_inAlu: the variant inside alu element
  + chr: chromosome of a given variant
  + posi: chromosome position of a given variant
  + ntCh: nucleotide changes of a given variant
```{r add_info,include=TRUE,eval=FALSE}
rd_vcf_Alu <- rd[unique(queryHits(GenomicRanges::findOverlaps(rd,rd_Alu)))] # list of Alu elements
var_inAlu <- names(rd_vcf_Alu)
bigTab$inAlu <- as.integer(bigTab$variant %in% var_inAlu) # decide the given variants in Alu elements or not
bigTab$chr <- gsub("(.*):(.*)_(.*)","\\1",bigTab$variant) # extract chromosome
bigTab$position <- gsub("(.*):(.*)_(.*)","\\2",bigTab$variant) # extract position
bigTab$ntCh <- gsub("(.*):(.*)_(.*)","\\3",bigTab$variant) # nucleotide changes
```

## Separate samples with/without interferon treatment
- Separate samples with/without IFN treatment
- Merge with annotation
```{r sep_IFN,include=TRUE,eval=FALSE}
# select the samples without IFN
bigTab_ori <- bigTab[,c(names(bigTab)[!grepl(names(bigTab),pattern = "-2")])] # remove IFN samples
bigTab_ori <- merge(bigTab_ori,annoTab,id='variant',all.x=TRUE)
#
# retend IFN sample only
bigTab_IFN <- bigTab[,c('variant',names(bigTab)[grepl(names(bigTab),pattern = "-2")],'inAlu','ntCh')] 
bigTab_IFN <- merge(bigTab_IFN,annoTab,id='variant',all.x=TRUE)
```

## Identify and grouping candidates for samples without IFN treatment
- Select variants pass the cut-off defined in the beginning.
- Select the variants with AF >= AF_mean + 2 * AF_SD
- Grouping variants by the pattern of samples they were identified.
- Labeling variants with A/G or T/C and inside Alu elements
- Labeling the variants not in dbSNP
- Save results in a xlsx file
```{r ind_grp_woIFN,include=TRUE,eval=FALSE}
## Evaluate the DP/AC/ps passed the cut-off
tab_pf <- bigTab_ori
tab_pf$dpCut_C <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "DP_C")]]>=dp_cut)
tab_pf$dpCut_F <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "DP_F")]]>=dp_cut)
tab_pf$dpCut_X <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "DP_X")]]>=dp_cut)
tab_pf$dpCut_Z <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "DP_Z")]]>=dp_cut)
#
tab_pf$acCut_C <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "AC_C")]]>=ac_cut)
tab_pf$acCut_F <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "AC_F")]]>=ac_cut)
tab_pf$acCut_X <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "AC_X")]]>=ac_cut)
tab_pf$acCut_Z <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "AC_Z")]]>=ac_cut)
#
tab_pf$psCut_C <- as.integer(tab_pf$dpCut_C==3&tab_pf$acCut_C==num_cut)
tab_pf$psCut_F <- as.integer(tab_pf$dpCut_F==3&tab_pf$acCut_F==num_cut)
tab_pf$psCut_X <- as.integer(tab_pf$dpCut_X==3&tab_pf$acCut_X>=1) # either one pass the
tab_pf$psCut_Z <- as.integer(tab_pf$dpCut_Z==3&tab_pf$acCut_Z==num_cut)
#
tab_pf$psCut_CX <- as.integer(tab_pf$psCut_C==1&tab_pf$psCut_X==0)
tab_pf$psCut_FX <- as.integer(tab_pf$psCut_F==1&tab_pf$psCut_X==0)
tab_pf$psCut_ZX <- as.integer(tab_pf$psCut_Z==1&tab_pf$psCut_X==0)
#
tab_filt_ct <- tab_pf[tab_pf$psCut_CX==1|tab_pf$psCut_FX==1|tab_pf$psCut_ZX==1,]
#
# filter by AF changes, mean + 2 * SD
cL <- names(tab_filt_ct)[grepl(names(tab_filt_ct),pattern = "AF_C")]
fL <- names(tab_filt_ct)[grepl(names(tab_filt_ct),pattern = "AF_F")]
xL <- names(tab_filt_ct)[grepl(names(tab_filt_ct),pattern = "AF_X")]
zL <- names(tab_filt_ct)[grepl(names(tab_filt_ct),pattern = "AF_Z")]
#
for(k in 1:length(tab_filt_ct$variant)){
  cnum <- unlist(tab_filt_ct[k,cL])
  fnum <- unlist(tab_filt_ct[k,fL])
  xnum <- unlist(tab_filt_ct[k,xL])
  znum <- unlist(tab_filt_ct[k,zL])
  #
  if(sum(xnum)==0){cutX <- 0}else{cutX<-mean(xnum)+2*sd(xnum)}
  #
  tab_filt_ct$sigCX[k] <- as.integer(sum(as.integer(cnum>cutX))==3)
  tab_filt_ct$sigFX[k] <- as.integer(sum(as.integer(fnum>cutX))==3)
  tab_filt_ct$sigZX[k] <- as.integer(sum(as.integer(znum>cutX))==3)}
#
# grouping variants
tab_filt_ct_AF <- tab_filt_ct[tab_filt_ct$sigCX==1|tab_filt_ct$sigFX==1|tab_filt_ct$sigZX==1,]
#
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 1
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 1
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 1
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 1
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 2
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 2
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 2
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 3
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 3
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 3
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 4
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 5
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 6
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 7
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 8
#
tab_filt_ct_AF_inAlu <- rbind(tab_filt_ct_AF[tab_filt_ct_AF$inAlu==1&tab_filt_ct_AF$ntCh=="A/G",], # select A/G mutations and the position in Alu element
                              tab_filt_ct_AF[tab_filt_ct_AF$inAlu==1&tab_filt_ct_AF$ntCh=="T/C",]) # select T/C mutations and the position in Alu element
tab_filt_ct_AF_inAlu <- tab_filt_ct_AF_inAlu[!grepl(tab_filt_ct_AF_inAlu$SNP138,pattern = 'rs'),] #select variants not in SNP138 database
#
# Save results into xlsx file
rio::export(list("filtByCount"=tab_filt_ct,"filtByAF"=tab_filt_ct_AF,"inAlu"=tab_filt_ct_AF_inAlu),
            file = file.path(res_dir,"var_cand_nIFN.xlsx"))
```

## Identify and grouping candidates for samples with IFN treatment
- Select variants pass the cut-off defined in the beginning.
- Select the variants with AF >= AF_mean + 2 * AF_SD
- Grouping variants by the pattern of samples they were identified.
- Labeling variants with A/G or T/C and inside Alu elements
- Labeling the variants not in dbSNP
- Save results in a xlsx file
```{r ind_grp_wIFN,include=TRUE,eval=FALSE}
## Evaluate the DP/AC/ps passed the cut-off
tab_pf <- bigTab_IFN
tab_pf$dpCut_C <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "DP_C")]]>=dp_cut)
tab_pf$dpCut_F <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "DP_F")]]>=dp_cut)
tab_pf$dpCut_X <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "DP_X")]]>=dp_cut)
tab_pf$dpCut_Z <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "DP_Z")]]>=dp_cut)
#
tab_pf$acCut_C <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "AC_C")]]>=ac_cut)
tab_pf$acCut_F <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "AC_F")]]>=ac_cut)
tab_pf$acCut_X <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "AC_X")]]>=ac_cut)
tab_pf$acCut_Z <- rowSums(tab_pf[,names(tab_pf)[grepl(names(tab_pf),pattern = "AC_Z")]]>=ac_cut)
#
tab_pf$psCut_C <- as.integer(tab_pf$dpCut_C==3&tab_pf$acCut_C==3)
tab_pf$psCut_F <- as.integer(tab_pf$dpCut_F==3&tab_pf$acCut_F==3)
tab_pf$psCut_X <- as.integer(tab_pf$dpCut_X==3&tab_pf$acCut_X>=1) # either one pass the
tab_pf$psCut_Z <- as.integer(tab_pf$dpCut_Z==3&tab_pf$acCut_Z==3)
#
tab_pf$psCut_CX <- as.integer(tab_pf$psCut_C==1&tab_pf$psCut_X==0)
tab_pf$psCut_FX <- as.integer(tab_pf$psCut_F==1&tab_pf$psCut_X==0)
tab_pf$psCut_ZX <- as.integer(tab_pf$psCut_Z==1&tab_pf$psCut_X==0)
#
tab_filt_ct <- tab_pf[tab_pf$psCut_CX==1|tab_pf$psCut_FX==1|tab_pf$psCut_ZX==1,]
#
# filter by AF changes
cL <- names(tab_filt_ct)[grepl(names(tab_filt_ct),pattern = "AF_C")]
fL <- names(tab_filt_ct)[grepl(names(tab_filt_ct),pattern = "AF_F")]
xL <- names(tab_filt_ct)[grepl(names(tab_filt_ct),pattern = "AF_X")]
zL <- names(tab_filt_ct)[grepl(names(tab_filt_ct),pattern = "AF_Z")]
#
for(k in 1:length(tab_filt_ct$variant)){
  cnum <- unlist(tab_filt_ct[k,cL])
  fnum <- unlist(tab_filt_ct[k,fL])
  xnum <- unlist(tab_filt_ct[k,xL])
  znum <- unlist(tab_filt_ct[k,zL])
  #
  if(sum(xnum)==0){cutX <- 0}else{cutX<-mean(xnum)+2*sd(xnum)}
  #
  tab_filt_ct$sigCX[k] <- as.integer(sum(as.integer(cnum>cutX))==3)
  tab_filt_ct$sigFX[k] <- as.integer(sum(as.integer(fnum>cutX))==3)
  tab_filt_ct$sigZX[k] <- as.integer(sum(as.integer(znum>cutX))==3)}
#
# grouping variants
tab_filt_ct_AF <- tab_filt_ct[tab_filt_ct$sigCX==1|tab_filt_ct$sigFX==1|tab_filt_ct$sigZX==1,]
#
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 1
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 1
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 1
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 1
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 2
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 2
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 2
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 3
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 3
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 3
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 4
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 5
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&tab_filt_ct_AF$sigZX==1] <- 6
tab_filt_ct_AF$Grp[tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 7
tab_filt_ct_AF$Grp[!tab_filt_ct_AF$sigCX==1&!tab_filt_ct_AF$sigFX==1&!tab_filt_ct_AF$sigZX==1] <- 8
#
tab_filt_ct_AF_inAlu <- rbind(tab_filt_ct_AF[tab_filt_ct_AF$inAlu==1&tab_filt_ct_AF$ntCh=="A/G",], # select A/G mutations and the position in Alu element
                              tab_filt_ct_AF[tab_filt_ct_AF$inAlu==1&tab_filt_ct_AF$ntCh=="T/C",]) # select T/C mutations and the position in Alu element
tab_filt_ct_AF_inAlu <- tab_filt_ct_AF_inAlu[!grepl(tab_filt_ct_AF_inAlu$SNP138,pattern = 'rs'),] #select variants not in SNP138 database
#
rio::export(list("filtByCount"=tab_filt_ct,"filtByAF"=tab_filt_ct_AF,"inAlu"=tab_filt_ct_AF_inAlu),
            file = file.path(res_dir,"var_cand_IFN.xlsx"))
```

## Generate report with the editted Rmd file
- Input: The integrated information in xlsx file for samples with/without IFN treatment
- Process with the *report_rmd_v2.Rmd* file
```{r plot_rmd,include=TRUE,eval=FALSE}
# Copy the template Rmd file into main directory
file.copy(file.path(base_dir,"report_rmd_v2.Rmd"),res_dir,recursive = TRUE)
#
# Load the results of SNPs identified in the samples without IFN treatment
file_ori <- file.path(res_dir,"var_cand_nIFN.xlsx")
#
# Build report
rmarkdown::render(file.path(res_dir,"report_rmd_v2.Rmd"),
                  output_file = file.path(res_dir,"report_fileName_nonIFN.html"))
#
# load results of SNPs identified in the samples with IFN treatment
file_ori <- file.path(res_dir,"var_cand_IFN.xlsx")
#
# Build report
rmarkdown::render(file.path(res_dir,"report_rmd_v2.Rmd"),
                  output_file = file.path(res_dir,"report_fileName_IFN.html"))
```

## Sessions
```{r session,include=TRUE,eval=FALSE}
sessionInfo()
```

Bioinformatics Resource Center, The Rockefeller University