---
title: "RNASeq_varCall_Report_20190827"
author: "Ji-Dung Luo"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,cache = TRUE)
library(ggplot2)
library(DT)
library(plotly)
library(pheatmap)
library(seqLogo)
# library(ggpubr)
#
# load(envData)
```

## Filtering criteria
- variants with total depth >= 5 
- depth of mutant allele >= 2 in each sample (exclude fluc)
- variants present in all samples of each group
- not in dbSNP (hg19.snp138)

## Nucleotide conversions of the filtered variants
```{r tab_ntCh}
tabX <- rio::import(file_ori,which = "filtByAF")
tbl <- as.matrix(table(tabX$inAlu[!grepl(tabX$SNP138,pattern = 'rs')],
                       tabX$ntCh[!grepl(tabX$SNP138,pattern = 'rs')]))
rownames(tbl) <- c("nonAlu","inAlu")
tbl
#
matX <- data.frame(ntch=rep(colnames(tbl),2),
                   Alu=c(rep("nonAlu",length(colnames(tbl))),c(rep("inAlu",length(colnames(tbl))))),
                   count=c(tbl[1,],tbl[2,]))
#
p <- ggplot(matX,aes(ntch,count,fill=Alu))+geom_bar(stat='identity')+
  labs(x="",y="Variant counts",fill="")+theme_classic()
ggplotly(p)
```

## Functional categories of the filtered variants
```{r tab_gfunc}
tabX <- rio::import(file_ori,which = "filtByAF")
tbl <- as.matrix(table(tabX$inAlu[!grepl(tabX$SNP138,pattern = 'rs')],
                       tabX$geneFunc[!grepl(tabX$SNP138,pattern = 'rs')]))
rownames(tbl) <- c("nonAlu","inAlu")
tbl
#
matX <- data.frame(ntch=rep(colnames(tbl),2),
                   Alu=c(rep("nonAlu",length(colnames(tbl))),c(rep("inAlu",length(colnames(tbl))))),
                   count=c(tbl[1,],tbl[2,]))
#
p <- ggplot(matX,aes(ntch,count,fill=Alu))+geom_bar(stat='identity')+
  labs(x="",y="Variant counts",fill="")+theme_classic()+coord_flip()
ggplotly(p)
```

## Clustering of potential ADAR-edditing sites by groups
Groups from top to down: 1,2,3,4,5,6,7,8
```{r clustVar}
tabV <- rio::import(file_ori,which = "filtByAF")
tabV <- tabV[!grepl(tabV$SNP138,pattern = "rs"),]
tabV <- tabV[tabV$ntCh=="A/G"|tabV$ntCh=="T/C",]
tabV$inAlu <- c("FALSE","TRUE")[factor(tabV$inAlu,levels = c(0,1))]
tabV3 <- tabV[,c('variant',
                 names(tabV)[grepl(names(tabV),pattern='^AF_')],
                 names(tabV)[grepl(names(tabV),pattern='sig')],
                 "geneFunc","inAlu","Grp")]
tabV3 <- tabV3[,!grepl(names(tabV3),pattern = 'AF_X')]
#
tabV3$Grp <- factor(tabV3$Grp,levels = 1:8,ordered = TRUE)
tabV3 <- tabV3[order(tabV3$Grp),]
#
matV3 <- tabV3[,names(tabV3)[grepl(names(tabV3),pattern='^AF_')]]
rownames(matV3) <- tabV3$variant
colTab <- data.frame(Cell=c(rep("C",3),rep("F",3),rep("Y",3),rep("Z",3)))
rownames(colTab) <- colnames(matV3)
rowTab <- data.frame(type=tabV3$geneFunc,inAlu=tabV3$inAlu)
rownames(rowTab) <- tabV3$variant
#
tb <- table(tabV3$Grp)
gaps <- NULL
gaps <- unlist(tb[1])
for(k in 2:length(tb)){
  gaps[k] <- gaps[k-1]+tb[k]
}
set.seed(0)
pheatmap(matV3,cluster_cols = FALSE,cluster_rows = FALSE,
         color = colorRampPalette(c("white", "red"))(100),
         annotation_row = rowTab,annotation_col = colTab,
         show_rownames = FALSE,show_colnames = FALSE,gaps_row = gaps)
#
message("Types of mutations in different variants")
datatable(tabV3)
```

## Types of mutations in different groups of variants
```{r typ_grp}
tbl <- table(tabV3$Grp,tabV3$geneFunc)
tbl
tabH <- rbind(data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[1],length(colnames(tbl))),count=tbl[1,],percent=tbl[1,]/sum(tbl[1,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[2],length(colnames(tbl))),count=tbl[2,],percent=tbl[2,]/sum(tbl[2,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[3],length(colnames(tbl))),count=tbl[3,],percent=tbl[3,]/sum(tbl[3,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[4],length(colnames(tbl))),count=tbl[4,],percent=tbl[4,]/sum(tbl[4,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[5],length(colnames(tbl))),count=tbl[5,],percent=tbl[5,]/sum(tbl[5,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[6],length(colnames(tbl))),count=tbl[6,],percent=tbl[6,]/sum(tbl[6,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[7],length(colnames(tbl))),count=tbl[7,],percent=tbl[7,]/sum(tbl[7,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[8],length(colnames(tbl))),count=tbl[8,],percent=tbl[8,]/sum(tbl[8,])))
#
# tabH$type <- factor(tabH$type,levels = c("C/F/Y/Z","C/F/Y","C/F/Z","C/Y/Z","F/Y/Z",
#                                   "C/F","C/Y","C/Z","F/Y","F/Z","Y/Z","C","F","Y","Z"))
pc <- ggplot(tabH,aes(x=type,y=count,fill=Group))+geom_bar(stat='identity')+
  labs(x="",y="Counts",fill="Types")+
  scale_x_discrete(limits=unique(rev(tabH$type)))+
  theme_classic()+coord_flip()
pp <- ggplot(tabH,aes(x=type,y=percent,fill=Group))+geom_bar(stat='identity')+
  labs(x="",y="Proportion",fill="Types")+
  scale_x_discrete(limits=unique(rev(tabH$type)))+
  theme_classic()+coord_flip()
message("Counts of variant in different groups")
ggplotly(pc)
message("Proportion of variant in different groups")
ggplotly(pp)
```

## Potential edditing sites in Alu or outside of the Alu
```{r var_inAlu}
tbl <- table(tabV3$Grp,tabV3$inAlu)
tbl
tabH <- rbind(data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[1],length(colnames(tbl))),count=tbl[1,],percent=tbl[1,]/sum(tbl[1,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[2],length(colnames(tbl))),count=tbl[2,],percent=tbl[2,]/sum(tbl[2,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[3],length(colnames(tbl))),count=tbl[3,],percent=tbl[3,]/sum(tbl[3,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[4],length(colnames(tbl))),count=tbl[4,],percent=tbl[4,]/sum(tbl[4,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[5],length(colnames(tbl))),count=tbl[5,],percent=tbl[5,]/sum(tbl[5,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[6],length(colnames(tbl))),count=tbl[6,],percent=tbl[6,]/sum(tbl[6,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[7],length(colnames(tbl))),count=tbl[7,],percent=tbl[7,]/sum(tbl[7,])),
              data.frame(Group=colnames(tbl),type=rep(rownames(tbl)[8],length(colnames(tbl))),count=tbl[8,],percent=tbl[8,]/sum(tbl[8,])))
tabH$Group <- c("non-Alu","Alu")[as.integer(tabH$Group)]
#
pc <- ggplot(tabH,aes(x=type,y=count,fill=Group))+geom_bar(stat='identity')+
  labs(x="",y="Counts",fill="")+
  scale_x_discrete(limits=unique(rev(tabH$type)))+
  theme_classic()+coord_flip()
pp <- ggplot(tabH,aes(x=type,y=percent,fill=Group))+geom_bar(stat='identity')+
  labs(x="",y="Proportion",fill="")+
  scale_x_discrete(limits=unique(rev(tabH$type)))+
  theme_classic()+coord_flip()
message("Counts of variant in different groups")
ggplotly(pc)
message("Proportion of variant in different groups")
ggplotly(pp)
```

## Motif analysis
- select A/G variants
- sequence from -5 ~ +5 (central with the variant)
- seqLogo
```{r seqLogo}
tabV <- rio::import(file_ori,which = "filtByAF")
tabV <- tabV[!grepl(tabV$SNP138,pattern = "rs"),]
# tabV <- tabV[tabV$ntCh=="A/G"|tabV$ntCh=="T/C",]
tabV <- tabV[tabV$ntCh=="A/G",]
tabV$inAlu <- c("non-Alu","Alu")[factor(tabV$inAlu,levels = c(0,1))]
tabV3 <- tabV[,c('variant',
                 names(tabV)[grepl(names(tabV),pattern='AF_')],
                 names(tabV)[grepl(names(tabV),pattern='sig')],
                 "geneFunc","inAlu","Grp")]
#
message("All Groups in Alu")
varList <- tabV3$variant[tabV3$inAlu=="Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("All Groups not in Alu")
varList <- tabV3$variant[!tabV3$inAlu=="Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 1 in Alu")
varList <- tabV3$variant[tabV3$Grp=="1"&tabV3$inAlu=="Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 1 in nonAlu")
varList <- tabV3$variant[tabV3$Grp=="1"&tabV3$inAlu=="non-Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 2 in Alu")
varList <- tabV3$variant[tabV3$Grp=="2"&tabV3$inAlu=="Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 2 in nonAlu")
varList <- tabV3$variant[tabV3$Grp=="2"&tabV3$inAlu=="non-Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 3 in Alu")
varList <- tabV3$variant[tabV3$Grp=="3"&tabV3$inAlu=="Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 3 in nonAlu")
varList <- tabV3$variant[tabV3$Grp=="3"&tabV3$inAlu=="non-Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 4 in Alu")
varList <- tabV3$variant[tabV3$Grp=="4"&tabV3$inAlu=="Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 4 in nonAlu")
varList <- tabV3$variant[tabV3$Grp=="4"&tabV3$inAlu=="non-Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 5 in Alu")
varList <- tabV3$variant[tabV3$Grp=="5"&tabV3$inAlu=="Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 5 in nonAlu")
varList <- tabV3$variant[tabV3$Grp=="5"&tabV3$inAlu=="non-Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 6 in Alu")
varList <- tabV3$variant[tabV3$Grp=="6"&tabV3$inAlu=="Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 6 in nonAlu")
varList <- tabV3$variant[tabV3$Grp=="6"&tabV3$inAlu=="non-Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 7 in Alu")
varList <- tabV3$variant[tabV3$Grp=="7"&tabV3$inAlu=="Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 7 in nonAlu")
varList <- tabV3$variant[tabV3$Grp=="7"&tabV3$inAlu=="non-Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 8 in Alu")
varList <- tabV3$variant[tabV3$Grp=="8"&tabV3$inAlu=="Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
a <- seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
#
message("Group 8 in nonAlu")
varList <- tabV3$variant[tabV3$Grp=="8"&tabV3$inAlu=="non-Alu"]
rdSeq <- rd[names(rd)%in%varList]
ab <- consensusMatrix(rdSeq$seq)
ncTab <- ab[1:4,]
pcTab <- ncTab/colSums(ncTab)
pwTab <- makePWM(pcTab)
seqLogo(pwTab,ic.scale = FALSE)
seqLogo(pwTab,ic.scale = TRUE)
```

## Potential ADAR-edditing sites
only A/G (T/C) variations in Alu
```{r varTab}
tabV <- rio::import(file_ori,which = "inAlu")
tabV2 <- dplyr::select(tabV,variant,geneSymbol,geneFunc,exonFunc,AAChange,sigCX,sigFX,sigYX,sigZX,Grp)
datatable(tabV2)
```

Bioinformatics Resource Center, The Rockefeller University
